src/lib/structures/modelMap.ml

```diff
@@ -30,100 +30,182 @@
+          Fiber.pp fiber
+          Value.pp ret_val
+          aux seq
+    in
+    aux ppf (Value.Map.to_seq rel)
+end
 
-module V = Set.Make
+module P = Map.Make
     (struct
-      type t = (E.t * (X.r * string)) list * (X.r * string)
-      let compare (l1, (v1,_)) (l2, (v2,_)) =
-        let c = X.hash_cmp v1 v2 in
-        if c <> 0 then c
-        else
-          let c = ref 0 in
-          try
-            List.iter2
-              (fun (_,(x,_)) (_,(y,_)) ->
-                 let d = X.hash_cmp x y in
-                 if d <> 0 then begin c := d; raise Exit end
-              ) l1 l2;
-            !c
-          with
-          | Exit -> !c
-          | Invalid_argument _ -> List.length l1 - List.length l2
+      type t = sy [@@deriving ord]
     end)
 
-type key = P.key
-type elt = V.t
-
 type t = {
-  values : V.t P.t;
+  values : Graph.t P.t;
   suspicious : bool;
 }
 
-let is_suspicious_name hs =
-  match Hstring.view hs with
-  | "@/" | "@%" | "@*" -> true
-  | _ -> false
-
-(* The model generation is known to be imcomplete for FPA and Bitvector
-   theories. *)
-let is_suspicious_symbol = function
-  | Sy.Op (Float | Fixed | Abs_int | Abs_real | Sqrt_real
-          | Sqrt_real_default | Sqrt_real_excess
-          | Real_of_int | Int_floor | Int_ceil
-          | Max_int | Max_real | Min_int | Min_real
-          | Pow | Integer_log2 | Int2BV _ | BV2Nat
-          | BVand | BVor | Integer_round) -> true
-  | Sy.Name (hs, _, _) when is_suspicious_name hs -> true
-  | _ -> false
-
-let add ((sy, _, _) as p) v { values; suspicious } =
-  let prof_p = try P.find p values with Not_found -> V.empty in
-  let values =
-    if V.mem v prof_p then values
-    else P.add p (V.add v prof_p) values
-  in
-  { values; suspicious = suspicious || is_suspicious_symbol sy }
-
-let iter f { values; _ } =
-  P.iter (fun ((sy, _, _) as key) value ->
-      match sy with
-      | Sy.Name (_, _, true) ->
-        (* We don't print constants defined by the user. *)
-        ()
-      | _ -> f key value
-    ) values
-
-let[@inline always] is_suspicious { suspicious; _ } = suspicious
-
-let fold f { values; _ } acc = P.fold f values acc
-
-let empty = { values = P.empty; suspicious = false }
-
-let is_empty { values; _ } = P.is_empty values
+let add ((id, arg_tys, _) as sy) arg_vals ret_val { values; suspicious } =
+  if List.compare_lengths arg_tys arg_vals <> 0 then
+    Fmt.invalid_arg "The arity of the symbol %a doesn't agree the number of \
+                     arguments" Id.pp id;
+
+  let graph = try P.find sy values with Not_found -> Graph.empty in
+  let values = P.add sy (Graph.add arg_vals ret_val graph) values in
+  { values; suspicious }
+
+let empty ~suspicious = { values = P.empty; suspicious }
+
+let pp_named_arg_ty ~unused ppf (arg_name, arg_ty) =
+  let pp_unused ppf unused = if unused then Fmt.pf ppf "_" else () in
+  Fmt.pf ppf "(%aarg_%i %a)" pp_unused unused arg_name Ty.pp_smtlib arg_ty
```

Comment:
If we want to add `_` to the unused arguments, we should either do it properly or not do it. Currently we only add `_` to the arguments if they are *all* unused, which is a middle ground that I don't think makes much sense; I would prefer we did nothing here.

Reply:
Our expressions to print the graph have a very particular form. If we use one argument in our expression, we use all of them.